tests/client/test_auth.py

```diff
@@ -127,3 +200,142 @@ def test_auth_invalid_type():
     with Client(dispatch=MockDispatch(), auth="not a tuple, not a callable") as client:
         with pytest.raises(TypeError):
             client.get(url)
+
+
+def test_digest_auth_returns_no_auth_if_no_digest_header_in_response():
+    url = "https://example.org/"
+    auth = HTTPDigestAuth(username="tomchristie", password="password123")
+
+    with Client(dispatch=MockDispatch()) as client:
+        response = client.get(url, auth=auth)
+
+    assert response.status_code == 200
+    assert response.json() == {"auth": None}
+
+
+@pytest.mark.parametrize(
+    "algorithm,expected_hash_length,expected_response_length",
+    [
+        ("MD5", 64, 32),
+        ("MD5-SESS", 64, 32),
+        ("SHA", 64, 40),
+        ("SHA-SESS", 64, 40),
+        ("SHA-256", 64, 64),
+        ("SHA-256-SESS", 64, 64),
+        ("SHA-512", 64, 128),
+        ("SHA-512-SESS", 64, 128),
+    ],
+)
+def test_digest_auth(algorithm, expected_hash_length, expected_response_length):
+    url = "https://example.org/"
+    auth = HTTPDigestAuth(username="tomchristie", password="password123")
+
+    with Client(dispatch=MockDigestAuthDispatch(algorithm=algorithm)) as client:
+        response = client.get(url, auth=auth)
+
+    assert response.status_code == 200
+    auth = response.json()["auth"]
+    assert auth.startswith("Digest")
```

Comment:
Super nit-picky but we could do `auth.startswith("Digest ")` to ensure that the header is being formed correctly. Currently returning `WWW-Authenticate: DigestZ realm=...` would pass this test and below.

Reply:
That's not nit-picky at all :)

That scenario is covered in the malformed header test below. The `partition` call would return a `scheme` equal to `DigestZ` which the `if scheme.lower() != "digest"` guard would catch.

Reply:
I've added some more test cases including yours for sanity. :)